examples/microwave-oven-app/microwave-oven-common/include/microwave-oven-device.h

```diff
@@ -0,0 +1,235 @@
+     * @return Returns a CHIP_NO_ERROR if there was no error and the label was returned successfully.
+     * CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the modeIndex in beyond the list of available labels.
+     */
+    CHIP_ERROR GetModeLabelByIndex(uint8_t modeIndex, MutableCharSpan & label) override;
+    /**
+     * Get the mode value of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param value a reference to the uint8_t variable that is to contain the mode value.
+     * @return Returns a CHIP_NO_ERROR if there was no error and the value was returned successfully.
+     * CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the modeIndex in beyond the list of available values.
+     */
+    CHIP_ERROR GetModeValueByIndex(uint8_t modeIndex, uint8_t & value) override;
+
+    /**
+     * Get the mode tags of the Nth mode in the list of modes.
+     * @param modeIndex The index of the mode to be returned. It is assumed that modes are indexable from 0 and with no gaps.
+     * @param tags a reference to an existing and initialised buffer that is to contain the mode tags. std::copy can be used
+     * to copy into the buffer.
+     * @return Returns a CHIP_NO_ERROR if there was no error and the mode tags were returned successfully.
+     * CHIP_ERROR_PROVIDER_LIST_EXHAUSTED if the modeIndex in beyond the list of available mode tags.
+     */
+    CHIP_ERROR GetModeTagsByIndex(uint8_t modeIndex, DataModel::List<ModeTagStructType> & tags) override;
+
+private:
+    // define delegates and instances for Microwave Oven device
+    MicrowaveOvenControl::Instance mMicrowaveOvenControlInstance;
+    OperationalState::Instance mOperationalStateInstance;
+    ModeBase::Instance mMicrowaveOvenModeInstance;
+
+    // MicrowaveOvenMode types
+    using ModeTagStructType              = detail::Structs::ModeTagStruct::Type;
+    ModeTagStructType modeTagsNormal[1]  = { { .value = to_underlying(MicrowaveOvenMode::ModeTag::kNormal) } };
+    ModeTagStructType modeTagsDefrost[1] = { { .value = to_underlying(MicrowaveOvenMode::ModeTag::kDefrost) } };
+
+    const detail::Structs::ModeOptionStruct::Type kModeOptions[2] = {
+        detail::Structs::ModeOptionStruct::Type{ .label    = CharSpan::fromCharString("Normal"),
+                                                 .mode     = ModeNormal,
+                                                 .modeTags = DataModel::List<const ModeTagStructType>(modeTagsNormal) },
+        detail::Structs::ModeOptionStruct::Type{ .label    = CharSpan::fromCharString("Defrost"),
+                                                 .mode     = ModeDefrost,
+                                                 .modeTags = DataModel::List<const ModeTagStructType>(modeTagsDefrost) }
+    };
+
+    // Operational States
+    const OperationalState::GenericOperationalState opStateList[4] = {
```

Comment:
But should this be a static constexpr if this is not really endpoint-dependent?  No strong opinions on this part.